apps/omg/lib/omg/input_pointer.ex

```diff
@@ -14,34 +14,142 @@
 
 defmodule OMG.InputPointer do
   @moduledoc """
-  `OMG.InputPointer` and `OMG.InputPointer.Protocol` represent the data that's used to mention outputs that are intended
-  to be inputs to a transaction. Examples are UTXO positions or output id (consisting of transaction hash and output
-  index)
+  `OMG.InputPointer` represent the data that's used to mention outputs that are intended 
+  to be inputs to a transaction. Examples are UTXO positions or output id.
+  (consisting of transaction hash and output index)
+  """
+
+  # TODO(achiurizo)
+  # Do we really need this? How does this work with output type marker?
+  @input_pointer_type_marker <<1>>
+
+  @type utxo_pos_tuple() :: {
+          :utxo_position,
+          # blknum
+          non_neg_integer,
+          # txindex
+          non_neg_integer,
+          # oindex
+          non_neg_integer
+        }
+
+  # This is the 'legacy' db format that the code is using to maintain backwards
+  # compatibility.
+  @type utxo_pos_db_tuple() :: {
+          # blknum
+          non_neg_integer,
+          # txindex
+          non_neg_integer,
+          # oindex
+          non_neg_integer
+        }
+
+  # This tuple is the format used to push into the DB(RocksDB)
+  @type db_key_tuple() :: {
+          :input_pointer,
+          <<_::8>>,
+          utxo_pos_db_tuple()
+        }
+
+  @position_too_low_error_tuple {:error, :encoded_utxo_position_too_low}
+  @type position_too_low_error_tuple() :: {:error, :encoded_utxo_position_too_low}
+
+  defstruct [:blknum, :txindex, :oindex]
```

Comment:
I think a macro-wrapped `Utxo.position(x, y, z)` implementation was more readable and to-the-point. A Utxo position is naturally a triplet. We would consider maps, if we were considering expanding to new fields or any other modifications, but that is not the case. 

Also, I find `encode(Utxo.position(...))` more readable than `.pos(some_map_with_fields)`. The latter would also happily encode the position for anything _similar_ to a UTXO position, which is scary. 

I don't think this ever was unclear what `Utxo.position(...)` represents (as soon as someone knows what a UTXO position is) and it avoided the repetitive and noisy chatter `blknum: blknum, txindex: txindex` etc.

Reply:
>  think a macro-wrapped Utxo.position(x, y, z) implementation was more readable and to-the-point. A Utxo position is naturally a triplet. We would consider maps, if we were considering expanding to new fields or any other modifications, but that is not the case.

I think the underlying issue is we _should_ use maps instead. The tuple format while simple, doesn't really buy us much except cost us alot of code transition between a `:utxo_position` lead atom vs one with out (see the db key 
tuple).

> Also, I find encode(Utxo.position(...)) more readable than .pos(some_map_with_fields). The latter would also happily encode the position for anything similar to a UTXO position, which is scary.

I'm not sure the former really does much to prevent that either. You could [happily encode any value into the tuple as well](https://github.com/omisego/elixir-omg/blob/master/apps/omg/lib/omg/utxo.ex#L33). With the map, while more keys could be placed in there, the function pulls out the value it requires to build the utxo position, the remainder has no effect. the `ExPlasma.Utxo.pos/1` could also guard against valid values if needs be(it's already pattern matching the values out as a function, just need to include a guard).

> I don't think this ever was unclear what Utxo.position(...) represents (as soon as someone knows what a UTXO position is) and it avoided the repetitive and noisy chatter blknum: blknum, txindex: txindex etc.

The purpose of this removal was to move us away from using a tuple to represent this data. I think a map here provides much more clarity(and flexibility) when it comes to representing this data. Additionally, this PR wants to move towars restructuring how we organize this data. We go through a bit of effort to map a position just to point it back into the output utxo repreentation (see input utxo pos -> {output data, creating_tx_hash mapping).  I think we can simplify this and merge the input utxo information with that of its corresponding output utxo information. It only needs to be pulled out [individually when it comes to encoding](https://github.com/omisego/ex_plasma/blob/master/lib/ex_plasma/transaction.ex#L123).

I don't think it encourages that much noise, and even if the consensus deems it so, we could also do the same macro approach to produce the underlying map(instead of the current tuple).

Reply:
I will just add that, a macro-based-tripplet + macro guard validator is a very uncommon approach. From what I've seen, things were made more confusing than actually helping. Definitely for removal and usage of a common data structure like a struct.

Reply:
```
alias OMG.Utxo
require Utxo

Utxo.position(blknum, txindex, oindex)

# vs
%OMG.InputPointer{blknum: blknum, txindex: txindex, oindex: oindex}
```

Okay, let's be more fair. You need the `require` to invoke the macro.

> I'm surprised about this being confusing and the (small) uncommonness being a problem. I think what is built on top (MoreVP/State logic/Watcher logic/challenges/whatnot) is uncommon and confusing, and we could use this syntax sugar to make the higher-level constructs less confusing.

I am all for syntactic sugar to make code nicer. I think having a struct here can increase clarity and flexibility in our code then having a macro that returns us a atom led tuple that we convert back and forth into, a non atom-led tuple:

```
{:utxo_position, 1, 2, 3}
{1, 2, 3}
```